File: PlayWriter/Practical/Advance/playwright-pom-practice/pages/LoginPage.ts
import { Locator, Page } from '@playwright/test';

export class LoginPage {
  readonly page: Page;
  readonly emailInput: Locator;
  readonly passwordInput: Locator;
  readonly loginButton: Locator;
  readonly errorMessage: Locator;
  readonly successMessage: Locator;

  constructor(page: Page) {
    this.page = page;
    this.emailInput = page.getByLabel('Login Email');
    this.passwordInput = page.getByLabel('Login Password');
    this.loginButton = page.getByRole('button', { name: 'Login' });
    this.errorMessage = page.locator('#login-error');
    this.successMessage = page.locator('#login-success');
  }

  async open() {
    await this.page.goto('/');
  }

  async enterEmail(email: string) {
    await this.emailInput.fill(email);
  }

  async enterPassword(password: string) {
    await this.passwordInput.fill(password);
  }

  async submit() {
    await this.loginButton.click();
  }

  async login(email: string, password: string) {
    await this.enterEmail(email);
    await this.enterPassword(password);
    await this.submit();
  }
}

File: ManualTransition/Week2 Selenium Rebuild Properly/Practical/mini-selenium-framework-lab-practice/src/main/java/com/techlambda/selenium/miniframework/pages/LoginPage.java
package com.techlambda.selenium.miniframework.pages;

import com.techlambda.selenium.miniframework.utils.WaitUtility;
import org.openqa.selenium.By;
import org.openqa.selenium.WebDriver;

/**
 * LoginPage is part of the Page Object Model layer.
 * It contains login page locators and login page actions only.
 */
public class LoginPage {

    private final WebDriver driver;
    private final WaitUtility wait;

    private final By emailInput = By.id("email");
    private final By passwordInput = By.id("password");
    private final By loginButton = By.id("login-button");
    private final By loginError = By.id("login-error");

    public LoginPage(WebDriver driver, WaitUtility wait) {
        this.driver = driver;
        this.wait = wait;
    }

    public void enterEmail(String email) {
        wait.waitForVisible(emailInput).clear();
        driver.findElement(emailInput).sendKeys(email);
    }

    public void enterPassword(String password) {
        wait.waitForVisible(passwordInput).clear();
        driver.findElement(passwordInput).sendKeys(password);
    }

    public void clickLogin() {
        wait.waitForClickable(loginButton).click();
    }

    public void login(String email, String password) {
        enterEmail(email);
        enterPassword(password);
        clickLogin();
    }

    public String getErrorMessage() {
        return wait.waitForVisible(loginError).getText();
    }
}
